keras/utils/io_utils.py

```diff
@@ -155,3 +164,185 @@ def ask_to_proceed_with_overwrite(filepath):
+
+            if val and sys.version_info[0] == 3 and isinstance(val[0], str):
+                # convert to bytes
+                val = [x.encode('utf-8') for x in val]
+
+            data_npy = np.asarray(val)
+
+            num_chunks = 1
+            chunked_data = np.array_split(data_npy, num_chunks)
+
+            # This will never loop forever thanks to the test above.
+            is_too_big = lambda x: x.nbytes > HDF5_OBJECT_HEADER_LIMIT
+            while any(map(is_too_big, chunked_data)):
+                num_chunks += 1
+                chunked_data = np.array_split(data_npy, num_chunks)
+
+            if num_chunks > 1:
+                for chunk_id, chunk_data in enumerate(chunked_data):
+                    self.data.attrs['%s%d' % (attr, chunk_id)] = chunk_data
+            else:
+                self.data.attrs[attr] = val
+        else:
+            self.data.attrs[attr] = val
+
+    def __getitem__(self, attr):
+        if isinstance(self.data, dict):
+            if isinstance(attr, bytes):
+                attr = attr.decode('utf-8')
+            if attr in self.data:
+                val = self.data[attr]
+                if isinstance(val, dict) and val.get('_is_group'):
+                    val = H5Dict(val)
+                elif '_{}_pickled'.format(attr) in self.data:
+                    val = pickle.loads(val)
+                return val
+            else:
+                if self.read_only:
+                    raise ValueError('Can not create group in read only mode.')
+                val = {'_is_group': True}
+                self.data[attr] = val
+                return H5Dict(val)
+        if attr in self.data.attrs:
+            val = self.data.attrs[attr]
+            if type(val).__module__ == np.__name__:
+                if val.dtype.type == np.string_:
+                    val = val.tolist()
+        elif attr in self.data:
+            val = self.data[attr]
+            if isinstance(val, h5py.Dataset):
+                val = np.asarray(val)
+            else:
+                val = H5Dict(val)
+        else:
+            # could be chunked
+            chunk_attr = '%s%d' % (attr, 0)
+            is_chunked = chunk_attr in self.data.attrs
+            if is_chunked:
+                val = []
+                chunk_id = 0
+                while chunk_attr in self.data.attrs:
+                    chunk = self.data.attrs[chunk_attr]
+                    val.extend([x.decode('utf8') for x in chunk])
+                    chunk_id += 1
+                    chunk_attr = '%s%d' % (attr, chunk_id)
+            else:
+                if self.read_only:
+                    raise ValueError('Can not create group in read only mode.')
+                val = H5Dict(self.data.create_group(attr))
+        return val
+
+    def __len__(self):
+        return len(self.data)
+
+    def __iter__(self):
+        return iter(list(self.data))
+
+    def iter(self):
+        return list(self.data)
```

Comment:
self.data could get quite big, shouldn't we just do 'self.data.items()' to be more efficient?